src/blocks/privacy/v4l.rs

```diff
@@ -0,0 +1,155 @@
+        );
+
+        let mut s = Self {
+            config,
+            devices: HashMap::new(),
+            interval: interval(duration),
+            watches,
+            updates,
+        };
+        s.update_devices().await?;
+
+        Ok(s)
+    }
+
+    async fn update_devices(&mut self) -> Result<bool> {
+        let mut changes = false;
+        let mut devices_to_remove: HashMap<PathBuf, WatchDescriptor> = self.devices.clone();
+        let mut sysfs_paths = read_dir(V4L_DEVICES_PATH)
+            .await
+            .error("Unable to read /dev/v4l/by-id")?;
+        while let Some(sysfs_path) = sysfs_paths
+            .next_entry()
+            .await
+            .error("Unable to get next device in /dev/v4l/by-id")?
+        {
+            let sysfs_path = sysfs_path.path();
+            let link_path = PathBuf::from(V4L_DEVICES_PATH)
+                .join(sysfs_path.read_link().error("Could not read link")?)
+                .canonicalize()
+                .error("Failed to canonicalize path")?;
+
+            if self.config.exclude_device.contains(&link_path) {
+                debug!("ignoring {:?}", link_path);
+                continue;
+            }
+
+            if self.devices.contains_key(&link_path) {
+                devices_to_remove.remove(&link_path);
+            } else {
+                debug!("adding watch {:?}", link_path);
+                self.devices.insert(
+                    link_path.clone(),
+                    self.watches
+                        .add(&link_path, WatchMask::OPEN | WatchMask::CLOSE)
+                        .error("Failed to watch data location")?,
+                );
+                changes = true;
+            }
+        }
+        for (sysfs_path, wd) in devices_to_remove {
+            debug!("removing watch {:?}", sysfs_path);
+            self.devices.remove(&sysfs_path);
+            self.watches
+                .remove(wd)
+                .error("Failed to unwatch data location")?;
+            changes = true;
+        }
+
+        Ok(changes)
+    }
+}
+
+#[async_trait]
+impl<'a> PrivacyMonitor for Monitor<'a> {
+    async fn get_info(&mut self) -> Result<PrivacyInfo> {
+        let mut mapping: PrivacyInfo = PrivacyInfo::new();
+
+        let mut proc_paths = read_dir("/proc").await.error("Unable to read /proc")?;
+        while let Some(proc_path) = proc_paths
+            .next_entry()
+            .await
+            .error("Unable to get next device in /proc")?
+        {
+            let proc_path = proc_path.path();
+            let fd_path = proc_path.join("fd");
+            if let Ok(mut fd_paths) = read_dir(fd_path).await {
+                while let Ok(Some(fd_path)) = fd_paths.next_entry().await {
+                    if let Ok(link_path) = fd_path.path().read_link() {
+                        if self.devices.contains_key(&link_path) {
+                            if let Ok(mut file) = File::open(proc_path.join("comm")).await {
```

Comment:
Let's use `let Ok(...) = ... else { continue };` to avoid this level of nesting :)